### DISCUSSION TITLE
Few questions regarding the routing mechanism of shadow

### DISCUSSION THREAD

**Discussion by michaelmirkin** [shadow/shadow#692] [Q&A] [answered]
Hi,
I'm new to shadow and I'm trying to use shadow to simulate gossip protocol among a large number of nodes.

I've noticed a few strange phenomenons that potentially caused by incorrect use of shadow or alternatively by some bug (wither in my program or in shadow itself).  

As first step, I used the default topology (at ~/.shadow/share/topology.graphml.xml). I created a few hundreds of nodes and attaching each node to random geolocation. When the protocol starts, a gossip of a relatively large (1MB-10MB) piece of information begins. I noticed that in rare cases (~1/100 of nodes), stop getting some information after the sending interaction initiated. The sending node "believes" that the send was completed and the receiving node just stops getting EPOLLIN events. After debugging the issue, I have reasons to believe that this happens due to some kind of congestion of packets in the routers.

Another issue is that I noticed that if the nodes are sending information while receiving information, the receiving rate becomes slower. 
I believe that this happens also due to congestion in the routers (POI). 

Which leads to a question whether I'm doing something wrong by using the default topology?
Should I maybe raise the bandwidth of POIs or do they have a buffer that I should increase?
I'm not sure that I completely understand the role of the POIs.

**Comment by robgjansen** [comment, marked as answer]
Sorry you are having issues!

I wonder if some of the issues you are observing is due to packet loss. Do you notice the same issues if you set packet loss to 0? (I assume you are using the latest version of Shadow.)

The asymmetric send/recv bandwidth issue may be due to packet overhead? You could analyze the log file lines containing `[shadow-heartbeat] [node]` to check whether the sending and receiving rates of the host network interface changes at the same time that your application throughput changes. If the issue occurs in both the application and in the network, that would seem like a bug?

The default topology is basically a toy example. We still include it in Shadow because it is small and for the most part allows experiments to function.

You might consider using a newer topology that I released as part of a 2018 research paper. See the `Shadow Network Configuration` section on this page:
https://tmodel-ccs2018.github.io/

The new topology file itself is here:
https://tmodel-ccs2018.github.io/data/shadow/network/atlas.201801.shadow113.graphml.xml.xz

**Comment by michaelmirkin** [comment]
Thank you for your answer!

> I wonder if some of the issues you are observing are due to packet loss. Do you notice the same issues if you set packet loss to 0? (I assume you are using the latest version of Shadow.)

Yes. I did it and it didn't help.



> The asymmetric send/recv bandwidth issue may be due to packet overhead? You could analyze the log file lines containing `[shadow-heartbeat] [node]` to check whether the sending and receiving rates of the host network interface changes at the same time that your application throughput changes. If the issue occurs in both the application and in the network, that would seem like a bug?

After following your instructions I indeed see something weird:
```
00:01:20.435535 [thread-0] 00:01:17.000000000 [message] [client11~11.0.0.12] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,105468,2209040,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;1598,105468,1598,105468,0,0,0,0,0,0,0,0;1541,2209040,0,0,0,0,1322,87252,1820688,219,14454,286646
00:01:23.402223 [thread-0] 00:01:18.000000000 [message] [client10~11.0.0.11] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,2239172,101772,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;1538,2239172,0,0,0,0,893,58938,1251921,645,42570,885743;1542,101772,1542,101772,0,0,0,0,0,0,0,0
00:01:23.402232 [thread-0] 00:01:18.000000000 [message] [client11~11.0.0.12] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,95040,2117002,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;1440,95040,1440,95040,0,0,0,0,0,0,0,0;1455,2117002,0,0,0,0,807,53262,1130927,648,42768,890045
00:01:25.944333 [thread-0] 00:01:19.000000000 [message] [client10~11.0.0.11] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,1500,132,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;1,1500,0,0,0,0,0,0,0,1,66,1434;2,132,2,132,0,0,0,0,0,0,0,0
00:01:25.944344 [thread-0] 00:01:19.000000000 [message] [client11~11.0.0.12] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,6798,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;103,6798,103,6798,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:01:28.609656 [thread-0] 00:01:20.000000000 [message] [client10~11.0.0.11] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:01:28.609666 [thread-0] 00:01:20.000000000 [message] [client11~11.0.0.12] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
```
client11 sends data **only** to client10 (at this point of time), it seems judging by my app log and the quoted shadow log that some data is lost in the network. Do you have a clue of what can cause this?



> You might consider using a newer topology that I released as part of a 2018 research paper. See the `Shadow Network Configuration` section on this page:
> https://tmodel-ccs2018.github.io/
> 
> The new topology file itself is here:
> https://tmodel-ccs2018.github.io/data/shadow/network/atlas.201801.shadow113.graphml.xml.xz

Thank you for that! It is exactly what I need.

**Comment by robgjansen** [comment]
> Thank you for that! It is exactly what I need.

Did that new topology file fix your issues? I hope it did!

Otherwise, I don't have any solid ideas about why data appears to be getting "dropped" or "stuck" in the network. It could be that the packets are lost, but I think the more likely reason is that the data is there but the epoll mechanism did not notice the data and never told the application it needs to read. In order to understand that, we would have to debug the socket buffer on the receiver to see if it is empty or not at the point when things get stuck.

I wonder if your application is using TCP or UDP? Or is it doing anything else strange or non-standard?

**Comment by robgjansen** [comment]
Please re-open if you are still experiencing issues.
